enterprise/server/webhooks/github/github.go

```diff
@@ -0,0 +1,127 @@
+
+	gh "github.com/google/go-github/github"
+)
+
+func ParseRequest(r *http.Request) (*webhook_data.WebhookData, error) {
+	payload, err := webhookJSONPayload(r)
+	if err != nil {
+		return nil, status.InvalidArgumentErrorf("failed to parse webhook payload: %s", err)
+	}
+
+	// Uncomment this line to log the request payload (useful for getting more unit test data):
+	// fmt.Printf("=== PAYLOAD ===\n%s\n=== END OF PAYLOAD ===\n", string(payload))
+
+	event, err := gh.ParseWebHook(gh.WebHookType(r), payload)
+	if err != nil {
+		return nil, status.InvalidArgumentErrorf("failed to parse webhook payload: %s", err)
+	}
+	log.Printf("Handling GitHub webhook event: %T\n", event)
+	switch event := event.(type) {
+	case *gh.PullRequestEvent:
+		v, err := fieldValues(event, "Action", "PullRequest.Head.SHA", "PullRequest.Head.Repo.CloneURL")
+		if err != nil {
+			return nil, err
+		}
+		// Only build when the PR is opened or pushed to.
+		if !(v["Action"] == "opened" || v["Action"] == "synchronized") {
+			return nil, nil
+		}
+		return &webhook_data.WebhookData{
+			Repo: v["PullRequest.Head.Repo.CloneURL"],
+			SHA:  v["PullRequest.Head.SHA"],
+		}, nil
+	case *gh.PushEvent:
+		v, err := fieldValues(event, "Ref", "HeadCommit.ID", "Repo.CloneURL", "Repo.DefaultBranch")
+		if err != nil {
+			return nil, err
+		}
+		defaultBranchRef := fmt.Sprintf("refs/heads/%s", v["Repo.DefaultBranch"])
+		// Only run workflows on the default branch.
+		if v["Ref"] != defaultBranchRef {
+			return nil, nil
+		}
+		return &webhook_data.WebhookData{
+			Repo: v["Repo.CloneURL"],
+			// For some reason, "HeadCommit.SHA" is nil, but ID has the commit SHA,
+			// so we use that instead.
+			SHA: v["HeadCommit.ID"],
+		}, nil
+	}
+	return nil, nil
+}
+
+func webhookJSONPayload(r *http.Request) ([]byte, error) {
+	contentType, _, err := mime.ParseMediaType(r.Header.Get("content-type"))
+	if err != nil {
+		return nil, status.InvalidArgumentErrorf("failed to parse content type: %s", err)
+	}
+	switch contentType {
+	case "application/json":
+		body, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			return nil, status.InternalErrorf("failed to read request body: %s", err)
+		}
+		return body, nil
+	case "application/x-www-form-urlencoded":
+		body, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			return nil, status.InternalErrorf("failed to read request body: %s", err)
+		}
+		form, err := url.ParseQuery(string(body))
+		if err != nil {
+			return nil, err
+		}
+		const payloadFormParam = "payload"
+		return []byte(form.Get(payloadFormParam)), nil
+	default:
+		return nil, status.InvalidArgumentErrorf("unhandled MIME type: %s", contentType)
+	}
+}
+
+// fieldValues extracts values from an object given field paths which may include dot
+// separators. As the paths are traversed, pointers will be de-referenced. If a nil
+// value is encountered in any path, an error is returned.
+func fieldValues(obj interface{}, fieldPaths ...string) (map[string]string, error) {
```

Comment:
wrote this because I got sick of writing stuff like

```go
repo := event.Repo
if repo == nil {
  return nil, nilFieldError("Repo")
}
defaultBranch := repo.DefaultBranch
if defaultBranch == nil {
  return nil, nilFieldError("Repo.DefaultBranch")
}
...
```

and I figured we can reuse this for BitBucket and GitLab. not super attached to it though so LMK if you think it's not worth the extra code.

Reply:
OK, sgtm.